src/dmd/expressionsem.d

```diff
@@ -8558,6 +8558,17 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
         Lnomatch:
         }
 
+        if (exp.op == TOK.assign &&
+            exp.e1.op == TOK.variable &&
+            exp.e2.op == TOK.variable)
+        {
+            if ((cast(VarExp)exp.e1).var.ident is
```

Comment:
You must use `(cast(VarExp)exp.e1).var.equals((cast(VarExp)exp.e1).var)`.
Also this does not cover all the self-assignment cases. Things like `*sym = *sym`, or when aliases are used and probably others.

Reply:
What's the difference between using `equals` and `==`? I wonder because your proposed alternative fails in the same way as

```
src/dmd/iasmgcc.d(404): Deprecation: self-assignment of variables is not allowed
```

@NilsLankila, do you have any advice on how to perform equality check in the general case? Is this related to the return value of `expressionSemantic`?

Reply:
The implementation of `__traits(isSame, ....)` might help:

https://github.com/dlang/dmd/blob/04bf3325a3855f622d047d9317dd7a64a70f3956/src/dmd/traits.d#L1960

Reply:
The original problem was that you were comparing Identifiers, which are non-unique (Identifier.idPool has an internal hash table).